apps/gateway/src/middlewares/models/gemini.ts

```diff
@@ -0,0 +1,218 @@
+
+import { ModelAdapterBase } from "./model";
+
+import type { ProviderOptions } from "@ai-sdk/provider-utils";
+
+export abstract class GeminiModelAdapter extends ModelAdapterBase {
+  readonly modality = "chat";
+  readonly owned_by = "google";
+  readonly pricing = {
+    monthly_free_tokens: 0,
+  };
+
+  transformOptions(options?: ProviderOptions): ProviderOptions {
+    const { openaiCompatible: openAiOptions, ...rest } = options || {};
+
+    if (!openAiOptions) return rest;
+
+    const config: Record<string, any> = {};
+    const reasoning = (openAiOptions as any)
+      ?.reasoning as OpenAICompatibleReasoning;
+
+    if (reasoning) {
+      const thinkingConfig = this.transformReasoning(reasoning);
+      if (thinkingConfig) {
+        config.thinkingConfig = thinkingConfig;
+      }
+    }
+
+    return {
+      ...rest,
+      openaiCompatible: config,
+    };
+  }
+
+  protected transformReasoning(
+    params: OpenAICompatibleReasoning,
+  ): Record<string, any> | undefined {
+    const isReasoningActive =
+      (params.enabled === true ||
+        (params.enabled === undefined &&
+          (params.max_tokens !== undefined || params.effort !== undefined))) &&
+      params.effort !== "none";
+
+    if (isReasoningActive) {
+      const thinkingConfig: Record<string, any> = {};
+
+      thinkingConfig.includeThoughts =
+        params.enabled !== false && params.exclude !== true;
+
+      const specificConfig = this.getThinkingConfig(params);
+      Object.assign(thinkingConfig, specificConfig);
+
+      return thinkingConfig;
+    }
+
+    return undefined;
+  }
+
+  protected getThinkingConfig(
+    params: OpenAICompatibleReasoning,
+  ): Record<string, any> {
+    const thinkingConfig: Record<string, any> = {};
+
+    if (params.max_tokens === undefined) {
+      switch (params.effort) {
+        case "minimal":
+        case "low": {
+          thinkingConfig.thinkingBudget = 1024;
+          break;
+        }
+        case "high": {
+          thinkingConfig.thinkingBudget = 24_576;
+          break;
+        }
+        case "xhigh": {
+          thinkingConfig.thinkingBudget = 32_768; // assuming
+          break;
+        }
+        default: {
+          thinkingConfig.thinkingBudget = 8192;
+          break;
+        }
+      }
+    } else {
+      thinkingConfig.thinkingBudget = params.max_tokens;
+    }
+
+    return thinkingConfig;
+  }
+}
+
+export abstract class Gemini3ModelAdapter extends GeminiModelAdapter {
+  transformPrompt(prompt: any): any {
+    return prompt.map((message: any) => {
+      if (message.role === "assistant" && Array.isArray(message.content)) {
+        let hasInjected = false;
+        return {
+          ...message,
+          content: message.content.map((part: any) => {
+            if (part.type === "tool-call") {
+              if (hasInjected) {
+                return part;
+              }
+
+              const existingSignature = (part as any).providerOptions
```

Comment:
I don't think the signature is part of the providerOptions, is it? I thought it's part of the message parts instead?

Reply:
yes, it is part of message (tool call), but in separate field,  and since this is provider-specific (google), we use `providerOptions` for injection.